query-planner-js/src/composedSchema/buildComposedSchema.ts

```diff
@@ -0,0 +1,240 @@
+  const graphDirective = getJoinDirective('graph');
+
+  const graphEnumType = schema.getType(`${joinName}__Graph`);
+  assert(isEnumType(graphEnumType));
+
+  const graphMap: GraphMap = Object.create(null);
+
+  schema.extensions = {
+    ...schema.extensions,
+    federation: {
+      graphs: graphMap,
+    },
+  };
+
+  for (const graphValue of graphEnumType.getValues()) {
+    const name = graphValue.name;
+
+    const graphDirectiveArgs = getArgumentValuesForDirective(
+      graphDirective,
+      graphValue.astNode!,
+    );
+    assert(
+      graphDirectiveArgs,
+      `${graphEnumType.name} value ${name} in composed schema should have a @${graphDirective.name} directive`,
+    );
+
+    const serviceName: string = graphDirectiveArgs['name'];
+    const url: string = graphDirectiveArgs['url'];
+
+    graphMap[name] = {
+      serviceName,
+      url,
+    };
+  }
+
+  for (const type of Object.values(schema.getTypeMap())) {
+    if (isIntrospectionType(type)) continue;
+
+    // We currently only allow join spec directives on object types.
+    if (!isObjectType(type)) continue;
+
+    assert(
+      type.astNode,
+      `GraphQL type "${type.name}" should contain AST nodes`,
+    );
+
+    const ownerDirectiveArgs = getArgumentValuesForDirective(
+      ownerDirective,
+      type.astNode,
+    );
+
+    const typeMetadata: FederationTypeMetadata = ownerDirectiveArgs
+      ? {
+          serviceName: graphMap[ownerDirectiveArgs?.['graph']].serviceName,
+          keys: new MultiMap(),
+          isValueType: false,
+        }
+      : {
+          isValueType: true,
+        };
+
+    type.extensions = {
+      ...type.extensions,
+      federation: typeMetadata,
+    };
+
+    const typeDirectivesArgs = getArgumentValuesForRepeatableDirective(
+      typeDirective,
+      type.astNode,
+    );
+
+    assert(
+      !(typeMetadata.isValueType && typeDirectivesArgs.length >= 1),
+      `GraphQL type "${type.name}" cannot have a @${typeDirective.name} \
+directive without an $${ownerDirective.name} directive`,
+    );
+
+    for (const typeDirectiveArgs of typeDirectivesArgs) {
+      const serviceName = graphMap[typeDirectiveArgs['graph']].serviceName;
+
+      const keyFields = parseFieldSet(typeDirectiveArgs['key']);
+
+      typeMetadata.keys?.add(serviceName, keyFields);
+    }
+
+    for (const fieldDef of Object.values(type.getFields())) {
+      assert(
+        fieldDef.astNode,
+        `Field "${type.name}.${fieldDef.name}" should contain AST nodes`,
+      );
+
+      const fieldDirectiveArgs = getArgumentValuesForDirective(
+        fieldDirective,
+        fieldDef.astNode,
+      );
+
+      if (!fieldDirectiveArgs) continue;
+
+      const fieldMetadata: FederationFieldMetadata = {
+        serviceName: graphMap[fieldDirectiveArgs?.['graph']]?.serviceName,
+      };
+
+      fieldDef.extensions = {
+        ...fieldDef.extensions,
+        federation: fieldMetadata,
+      };
+
+      if (fieldDirectiveArgs) {
```

Comment:
This condition is always true, due to a continue above. Is that because the `continue` shouldn't be there or because this can get un-nested?